core/src/types.ts

```diff
@@ -152,3 +152,41 @@ export declare interface CollateralStatus {
     address?: string | null;
     balance?: BigNumber;
 }
+
+export type SurplusAuctionStates =
+    | 'just-started'
+    | 'have-bids'
+    | 'ready-for-collection'
+    | 'collected'
+    | 'requires-restart';
+
+export type SurplusAuctionEventTypes = 'start' | 'bid' | 'collect';
+
+export declare interface SurplusAuctionData {
+    bidAmountMKR?: BigNumber;
+    receiveAmountDAI: BigNumber;
+    receiverAddress: string;
+    auctionEndDate: Date;
+    bidEndDate?: Date;
+    earliestEndDate: Date;
+    state: SurplusAuctionStates;
+}
+
+export declare interface SurplusEvent extends SurplusAuctionData {
+    type: SurplusAuctionEventTypes;
+    address: string;
+    transactionHash: string;
+    transactionDate: Date;
+}
+
+export declare interface InitialSurplusAuction extends SurplusAuctionData {
+    id: number;
+    network: string;
+    events: SurplusEvent[];
+}
+
+export declare interface SurplusAuction extends InitialSurplusAuction {
+    marketUnitPrice: BigNumber;
+    highestBid?: BigNumber;
+    auctionPrice?: BigNumber;
+}
```

Comment:
regarding the hierarchy:

- i don't really understand why there's a duplicative information in various places:
  - the event has the auction data. although i would expect it to be the other way round: the events happen for the particular auction, so having the events listed as part of auction data makes more sense to me from the generic perspective.
  - if i think about it extensively: i fail to see the purpose of the initial auction here. Moreover i only can vaguely guess what "initial" implies in the naming.

Reply:
> * the event has the auction data. although i would expect it to be the other way round: the events happen for the particular auction, so having the events listed as part of auction data makes more sense to me from the generic perspective.

I took this directly from @valiafetisov type mocking (https://github.com/sidestream-tech/unified-auctions-ui/issues/188#issuecomment-1152161469). I understand your point, however I was thinking that maybe there was a reason behind this. But I agree it is generally not needed and could be removed.

> * if i think about it extensively: i fail to see the purpose of the initial auction here. Moreover i only can vaguely guess what "initial" implies in the naming.

I modeled this after how we currently calculate "normal auctions". Initial implies that is only contains the data directly fetched from the blockchain. The full object then adds any additional data, we calculate on our end. In this case we would calculate, who the last bidder was and then we set the highest bid and auction price according to this data. We also do this with the normal auctions, when we calculate the next price drop according to the step/cut parameters, or add the market value (which is not initially included and needs to be enriched afterwards).

Reply:
Verbal discussion result:

- dont use generics
- do not store auction data in the events
- store only relevant information in the events
- move the initial surplus aucion data pieces to the base interface: `id, network`

Rationales:

- most basic & simple, makes sense to not store auction data in the events.

Reply:
We agreed on the following types in call:

```js
import BigNumber from 'bignumber.js';

export declare interface GasParameters {
    maxFeePerGas?: string;
    maxPriorityFeePerGas?: string;
    gasPrice?: string;
}

export declare interface AuctionInitialInfo {
    network: string;
    id: string;
    index: number;
    collateralType: string;
    collateralSymbol: string;
    collateralAmount: BigNumber;
    debtDAI: BigNumber;
    startDate: Date;
    endDate: Date;
    vaultAddress: string;
    isActive: boolean;
    isFinished: boolean;
    isRestarting: boolean;
    marketUnitPrice?: BigNumber;
    marketUnitPriceToUnitPriceRatio?: BigNumber;
    transactionGrossProfit?: BigNumber;
    transactionAddress?: string;
    initialPrice: BigNumber;
    fetchedAt: Date;
}

export declare interface AuctionStatus {
    isActive: boolean;
    debtDAI: BigNumber;
    collateralAmount: BigNumber;
    unitPrice: BigNumber;
    totalPrice: BigNumber;
}

export declare interface Auction extends AuctionInitialInfo {
    unitPrice: BigNumber;
    totalPrice: BigNumber;
    approximateUnitPrice: BigNumber;
    secondsBetweenPriceDrops?: number;
    secondsTillNextPriceDrop?: number;
    priceDropRatio?: BigNumber;
    transactionGrossProfitDate?: Date;
    minimumBidDai: BigNumber;
}

export declare interface TransactionFees {
    bidTransactionFeeETH: BigNumber;
    bidTransactionFeeDAI: BigNumber;
    swapTransactionFeeETH: BigNumber;
    swapTransactionFeeDAI: BigNumber;
    authTransactionFeeETH: BigNumber;
    authTransactionFeeDAI: BigNumber;
    restartTransactionFeeETH: BigNumber;
    restartTransactionFeeDAI: BigNumber;
}

export declare interface CollateralRow extends CollateralConfig, Partial<MakerParams> {
    marketUnitPrice?: BigNumber | string;
    tokenAddress?: string;
    tokenAddressError?: string;
}

export declare interface AuctionTransaction extends Auction, TransactionFees {
    transactionNetProfit: BigNumber;
    combinedBidFeesDAI: BigNumber;
    combinedBidFeesETH: BigNumber;
    combinedSwapFeesDAI: BigNumber;
    combinedSwapFeesETH: BigNumber;
}

export declare interface RegularCalleeConfig {
    callee: 'UniswapV2CalleeDai' | 'WstETHCurveUniv3Callee' | 'CurveLpTokenUniv3Callee' | 'UniswapV3Callee';
    route: string[];
}

export declare interface UniswapV2LpTokenCalleeConfig {
    callee: 'UniswapV2LpTokenCalleeDai';
    token0: string;
    token1: string;
}

export declare interface CollateralConfig {
    title: string;
    ilk: string;
    symbol: string;
    decimals: number;
    exchange: RegularCalleeConfig | UniswapV2LpTokenCalleeConfig;
}

export declare interface NetworkConfig {
    chainId: string;
    title: string;
    url: string;
    gasPrice?: number;
    etherscanUrl: string;
    isFork: boolean;
}

export declare interface CalleeAddresses {
    UniswapV2CalleeDai: string;
    UniswapV2LpTokenCalleeDai: string;
    WstETHCurveUniv3Callee?: string;
    CurveLpTokenUniv3Callee?: string;
    UniswapV3Callee?: string;
}

export type CalleeNames = keyof CalleeAddresses;

export declare interface CalleeFunctions {
    getCalleeData: (network: string, collateral: CollateralConfig, profitAddress: string) => Promise<string>;
    getMarketPrice: (network: string, collateral: CollateralConfig, amount: BigNumber) => Promise<BigNumber>;
}

export declare interface MakerParams {
    secondsBetweenPriceDrops: number;
    priceDropRatio: BigNumber;
}

export declare interface MessageContent {
    key: string;
    content: string;
    duration: number;
}

export declare interface Notifier {
    (messageType: 'loading' | 'error' | 'success' | 'info', messageContent: MessageContent): void;
}

export declare interface TakeEvent {
    transactionHash: string;
    blockNumber: number;
    transactionDate?: Date;
}

export declare interface WalletBalances {
    walletETH: BigNumber;
    walletDAI: BigNumber;
    walletVatDAI: BigNumber;
    walletLastUpdatedDate: Date;
}

export declare interface CollateralStatus {
    type: string;
    symbol: string;
    isAuthorized: boolean;
    isAuthorizing: boolean;
    isDepositingOrWithdrawing: boolean;
    address?: string | null;
    balance?: BigNumber;
}

export type SurplusAuctionStates =
    | 'just-started'
    | 'have-bids'
    | 'ready-for-collection'
    | 'collected'
    | 'requires-restart';

export type SurplusAuctionEventTypes = 'start' | 'bid' | 'collect';

export declare interface SurplusAuctionData {
    id: number;
    network: string;
    receiveAmountDAI: BigNumber;
    receiverAddress: string;
    auctionEndDate: Date;
    bidEndDate?: Date;
    earliestEndDate: Date;
    state: SurplusAuctionStates;
}

export declare interface SurplusEvent {
    type: SurplusAuctionEventTypes;
    bidAmountMKR?: BigNumber;
    address: string;
    transactionHash: string;
    transactionDate: Date;
}

export declare interface InitialSurplusAuction extends SurplusAuctionData {
    events: SurplusEvent[];
}

export declare interface SurplusAuction extends InitialSurplusAuction {
    marketUnitPrice: BigNumber;
    highestBid?: BigNumber;
    auctionPrice?: BigNumber;
}

```

Reply:
>     receiverAddress: string;

NIT: I think the receiver address should be optional since for the state `just-started` (ie. auction without bids) we are not able to define it yet.

Reply:
> I think the receiver address should be optional since for the state just-started (ie. auction without bids) we are not able to define it yet.

Technically, in this case, `receiverAddress` would contain the address of the wallet that kicked the auction, so it doesn't need to be optional. But maybe there is a better name for this field

Reply:
> Technically, in this case, `receiverAddress` would contain the address of the wallet that kicked the auction, so it doesn't need to be optional. But maybe there is a better name for this field

As discussed in the Daily, we need to investigate the return values from the events. @valiafetisov will post any updates on the investigation. Until then the types will stay the same. Once we have more information we will adjust them accordingly.